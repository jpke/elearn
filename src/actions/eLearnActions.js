// import cookie from 'react-cookie'
import * as types from '../constants/actionTypes';

export function loading(item) {
  return {
    type: 'LOADING',
    item
  };
}

function loggedIn(response) {
  console.log("response: ", response);
  localStorage.setItem("login", {
    userName: response.name,
    user_Id: response._id,
    courses: response.courses,
    token: response.token
  });
  return {
    type: types.LOG_IN,
    userName: response.name,
    user_Id: response._id,
    courses: response.courses,
    token: response.token
  };
}

export function register(userName, email, password) {
  return function (dispatch) {
    dispatch(loading('register'));
    try {
      fetch('http://localhost:8080/elearn/users', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Accept': 'application/json'
        },
        body: JSON.stringify({
          name: userName,
          email: email,
          password: password
        })
      })
      .then(response => {
        if(response.status < 200 || response.status >= 300) {
          if(response.body.message === 'email already associated with an account') {
            return dispatch({
              type: types.EMAIL_USED,
              email
            });
          } else {
            let error = response;
            throw error;
          }
        }
        return response.json()
      })
      .then(response => {
        window.localStorage.loggedIn = true;
        dispatch(loading(''));
        dispatch(loggedIn(response));
      })
    } catch(error) {
      console.log("error response: ", error);
    }
  };
}

export function logIn(email, password) {
  return function (dispatch) {
    dispatch(loading('logIn'));
    try {
      fetch('http://localhost:8080/elearn/login', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Accept': 'application/json'
        },
        withCredentials: true,
        body: JSON.stringify({
          email: email,
          password: password
        })
      })
      .then(response => {
        if(response.status != 302) {
          let error = response;
          throw error;
        }
        return response.json()
      })
      .then(response => {
        window.localStorage.loggedIn = true;
        // console.log("set-cookie: ", response)
        // cookie.save("tokens",
        //   response.token,
        //   {
        //     httpOnly: true
        //   });
        dispatch(loading(''));
        dispatch(loggedIn(response));
      })
    } catch(error) {
      console.log("error response: ", error);
    }
  };
}

export function selectCourse(courseName, course_Id) {
  return {
    type: types.SELECT_COURSE,
    courseName,
    course_Id
  };
}

export function logOut() {
  //consider blacklisting token serverside in future
  window.localStorage.loggedIn = false;
  // cookie.remove('token');
  return {
    type: types.LOG_OUT
  };
}

export function startQuiz(token, course_Id) {
  return function (dispatch) {
    dispatch(loading('startQuiz'));
    //pull down quiz questions, then
    try {
      fetch('http://localhost:8080/elearn/quiz/'.concat(course_Id), {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${token}`
        }
      })
      .then(response => {
        if(response.status < 200 || response.status >= 300) {
          let error = response;
          throw error;
        }
        return response.json()
      })
      .then(response => {
        dispatch(loading(''));
        console.log("quiz actions: ", response);
        return dispatch({
          type: types.START_QUIZ,
          quizData: response
        });
        })
    } catch(error) {
      console.log("error response: ", error);
    }
  };
}

export function selectAnswer(answerSelected, idSelected, item) {
  return {
    type: types.SELECT_ANSWER,
    answerSelected,
    idSelected,
    itemSelected: parseInt(item)
  };
}

export function nextQuestion() {
  return {
    type: types.NEXT_QUESTION
  };
}

export function prevQuestion() {
  return {
    type: types.PREVIOUS_QUESTION
  };
}

export function submitQuiz(quizData, quizTitle, quizId, _id, token) {
  return function (dispatch) {
    dispatch(loading('submitQuiz'));
    try {
      fetch('http://localhost:8080/elearn/quiz/submit', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${token}`
        },
        body: JSON.stringify({
          title: quizTitle,
          quizData: quizData,
          quiz_Id: quizId,
          user_Id: _id
        })
      })
      .then(response => {
        if(response.status < 200 || response.status >= 300) {
          let error = response;
          throw error;
        } else {
          return response;
        }
      })
      .then(response => response.json())
      .then((response) => {
        console.log("action response: ", response);
        dispatch(loading(''));
        return dispatch({
          type: types.SUBMIT_QUIZ,
          score: response.score
        });
        })
    } catch(error) {
      console.log("error response: ", error);
    }
  }
}

export function getLessons(token) {
  return function (dispatch) {
    // let cookieToken = cookie.load('token');
    // console.log("cookieToken: ", cookieToken);
    dispatch(loading('lessons'));
    try {
      fetch('http://localhost:8080/elearn/lessons', {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${token}`
        },
        // withCredentials: true,
      })
      .then(response => {
        if(response.status < 200 || response.status >= 300) {
          let error = response;
          throw error;
        }
        return response.json()
      })
      .then(response => {
        dispatch(loading(''));
        return dispatch({
          type: types.GET_LESSONS,
          lessons: response.entries
        });
        })
    } catch(error) {
      console.log("error response: ", error);
    }
  };
}

export function getPDF(pdfId, token) {
  return function (dispatch) {
    dispatch(loading("pdf"));
    try {
      fetch('http://localhost:8080/elearn/lessons/'.concat(pdfId), {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${token}`
        },
        // withCredentials: true,
      })
      .then(response => {
        if(response.status < 200 || response.status >= 300) {
          let error = response;
          throw error;
        }
        return response.json()
      })
      .then(response => {
        dispatch(loading(""));
        console.log("response: ", response);
          return dispatch({
            type: types.GET_PDF,
            response
          });
        })
    } catch(error) {
      console.log("error response: ", error);
    }
  };
}
