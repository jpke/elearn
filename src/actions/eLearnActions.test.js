/* eslint-disable no-undef */
/* eslint-disable import/no-named-as-default */

// import * as types from '../constants/actionTypes';
import * as actions from './eLearnActions';

it('should create an action to update loading status upon async call', () => {
  expect(actions.loading("sample async call")).toMatchSnapshot();
})

it('should create an action to update loading status upon async completion', () => {
  expect(actions.loading("")).toMatchSnapshot();
})

it('should create an action to trigger error alert to user', () => {
  expect(actions.badResponse("sample problem occurred message")).toMatchSnapshot();
})

it('should create an action to update loading status upon async call', () => {
  expect(actions.loading("sample async call")).toMatchSnapshot();
})

// it('should create an action to call async registration endpoint', () => {
//   expect(actions.register()).toMatchSnapshot();
// })

it('should create an action to edit course title', () => {
  expect(actions.editCourse("000", "updated course title")).toMatchSnapshot();
})

// it('should create an action to call async endpoint adding user to a course', () => {
//   expect(actions.addUser()).toMatchSnapshot();
// })

// it('should create an action to call async endpoint deleting enrollable user from a course', () => {
//   expect(actions.deleteUser()).toMatchSnapshot();
// })

// it('should create an action to call async endpoint to update course', () => {
//   expect(actions.updateCourse()).toMatchSnapshot();
// })

// it('should create an action to call async endpoint logging user in', () => {
//   expect(actions.logIn()).toMatchSnapshot();
// })

it('should create an action to select a course from courses available to the user', () => {
  expect(actions.selectCourse("Selected Course")).toMatchSnapshot();
})

// it('should create an action to call async endpoint requesting users authorized for a course', () => {
//   expect(actions.getEnrollable()).toMatchSnapshot();
// })

// it('should create an action to call async endpoint removing enrolled user from a course', () => {
//   expect(actions.deleteUserFromCourse()).toMatchSnapshot();
// })

it('should create an action to log user out of client side', () => {
  expect(actions.logOut()).toMatchSnapshot();
})

// it('should create an action to call async endpoint returning certificate of completion of a course', () => {
//   expect(actions.getCertificate()).toMatchSnapshot();
// })

it('should create an action to create a new quiz on client side', () => {
  expect(actions.createQuiz()).toMatchSnapshot();
})

it('should create an action to edit selected quiz clientside', () => {
  expect(actions.editQuiz("1", "quiz item to edit", "updated value", "a")).toMatchSnapshot();
})

// it('should create an action to call async endpoint deleting quiz from database and course', () => {
//   expect(actions.deleteSavedQuiz()).toMatchSnapshot();
// })

it('should create an action to call async endpoint creating or updating a quiz in the database', () => {
  expect(actions.saveQuiz()).toMatchSnapshot();
})

// it('should create an action to select quiz from available quizzes and load quiz data from async endpoint', () => {
//   expect(actions.selectQuiz("some token", "quiz id", "user id")).toMatchSnapshot();
// })

it('should create an action to toggle view between list of available quizzes and quiz start or quiz edit page', () => {
  expect(actions.toggleQuizView()).toMatchSnapshot();
})

// it('should create an action to call async endpoint loading selecting quiz', () => {
//   expect(actions.loadQuiz()).toMatchSnapshot();
// })

it('should create an action to start quiz', () => {
  expect(actions.startQuiz()).toMatchSnapshot();
})

it('should create an action to record user selection of answer to quiz question', () => {
  expect(actions.selectAnswer("answer selected", "id of answer selected", "current quiz item")).toMatchSnapshot();
})

it('should create an action to advance to next question of quiz', () => {
  expect(actions.nextQuestion()).toMatchSnapshot();
})

it('should create an action to return to prior question of quiz', () => {
  expect(actions.prevQuestion()).toMatchSnapshot();
})

// it('should create an action to call async endpoint recording quiz submission', () => {
//   expect(actions.register()).toMatchSnapshot();
// })

// it('should create an action to call async endpoint retrieving course lesson list', () => {
//   expect(actions.getLessons()).toMatchSnapshot();
// })

it('should create an action to load link to view pdf lesson', () => {
  expect(actions.loadPreview("file data returned from api request")).toMatchSnapshot();
})

// it('should create an action to call async endpoint requesting url of pdf lesson preview', () => {
//   expect(actions.getPDF()).toMatchSnapshot();
// })

// it('should create an async action uploading pdf to course lessons', () => {
//   expect(actions.uploadPDF()).toMatchSnapshot();
// })

// it('should create an action to call async endpoint deleting pdf lesson from course lesson list', () => {
//   expect(actions.deletePDF()).toMatchSnapshot();
// })
