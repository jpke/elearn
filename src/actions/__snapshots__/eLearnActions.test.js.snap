exports[`test should create an action to advance to next question of quiz 1`] = `
Object {
  "type": "NEXT_QUESTION",
}
`;

exports[`test should create an action to call async endpoint creating or updating a quiz in the database 1`] = `[Function]`;

exports[`test should create an action to create a new quiz on client side 1`] = `
Object {
  "type": "CREATE_QUIZ",
}
`;

exports[`test should create an action to edit course title 1`] = `
Object {
  "id": "000",
  "type": "EDIT_COURSE",
  "value": "updated course title",
}
`;

exports[`test should create an action to edit selected quiz clientside 1`] = `
Object {
  "itemIndex": "1",
  "itemName": "quiz item to edit",
  "subIndex": "a",
  "type": "UPDATE_QUIZ",
  "value": "updated value",
}
`;

exports[`test should create an action to load link to view pdf lesson 1`] = `
Object {
  "selectedPdf": "",
  "type": "GET_PDF",
}
`;

exports[`test should create an action to log user out of client side 1`] = `
Object {
  "type": "LOG_OUT",
}
`;

exports[`test should create an action to record user selection of answer to quiz question 1`] = `
Object {
  "answerSelected": "answer selected",
  "idSelected": "id of answer selected",
  "itemSelected": NaN,
  "type": "SELECT_ANSWER",
}
`;

exports[`test should create an action to return to prior question of quiz 1`] = `
Object {
  "type": "PREVIOUS_QUESTION",
}
`;

exports[`test should create an action to select a course from courses available to the user 1`] = `
Object {
  "course": "Selected Course",
  "type": "SELECT_COURSE",
}
`;

exports[`test should create an action to start quiz 1`] = `
Object {
  "type": "START_QUIZ",
}
`;

exports[`test should create an action to toggle view between list of available quizzes and quiz start or quiz edit page 1`] = `
Object {
  "type": "TOGGLE_QUIZ_VIEW",
}
`;

exports[`test should create an action to trigger error alert to user 1`] = `
Object {
  "message": "sample problem occurred message",
  "type": "BAD_RESPONSE",
}
`;

exports[`test should create an action to update loading status upon async call 1`] = `
Object {
  "item": "sample async call",
  "type": "LOADING",
}
`;

exports[`test should create an action to update loading status upon async call 2`] = `
Object {
  "item": "sample async call",
  "type": "LOADING",
}
`;

exports[`test should create an action to update loading status upon async completion 1`] = `
Object {
  "item": "",
  "type": "LOADING",
}
`;
